tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -212,7 +229,431 @@
+      Operand<T> labels,
+      Operand<T> predictions,
+      Operand<TFloat32> thresholds,
+      Integer topK,
+      Integer classIndex,
+      Operand<T> sampleWeight,
+      boolean multiLabel,
+      Operand<T> labelWeights) {
+    if (multiLabel && labelWeights != null)
+      throw new IllegalArgumentException(
+          "labelWeights for multilabel data should be handled outside of updateConfusionMatrixVariables when multiLabel is true.");
+
+    if (variablesToUpdate == null || variablesToUpdate.isEmpty()) {
+      return Collections.EMPTY_LIST;
+    }
+
+    Operand<T> tLabels = labels;
+    Operand<T> tPredictions = predictions;
+    Operand<T> tSampleWeight = sampleWeight;
+
+    // size of the 0th dimension of thresholds
+    // reshape to scalar for operations later.
+    Operand<TInt32> numThresholds =
+        tf.reshape(tf.shape.size(thresholds, tf.constant(0)), tf.constant(Shape.scalar()));
+
+    // true if we will process thresholds as one-dimensional (possibly because we flatten them)
+    Operand<TBool> oneThresh;
+    if (multiLabel) {
+      oneThresh = tf.math.equal(tf.constant(1), tf.rank(thresholds));
+    } else {
+      // TODO handle Ragged Tensors????
+      // [y_pred,
+      //    y_true], _ = ragged_assert_compatible_and_get_flat_values([y_pred, y_true],
+      //                                                   sampleWeights)
+      oneThresh = tf.constant(true);
+    }
+
+    List<Op> controlOps = new ArrayList<>();
+    Operand<TInt32> axes = allAxes(tf, tPredictions);
+    controlOps.add(
+        tf.withSubScope("updateConfusionMatrixVariables-1")
+            .assertThat(
+                tf.reduceAll(
+                    tf.math.greaterEqual(
+                        tPredictions, cast(tf, tf.constant(0), tPredictions.type())),
+                    axes),
+                Collections.singletonList(tf.constant("predictions must be >= 0"))));
+    controlOps.add(
+        tf.withSubScope("updateConfusionMatrixVariables-2")
+            .assertThat(
+                tf.reduceAll(
+                    tf.math.lessEqual(tPredictions, cast(tf, tf.constant(1), tPredictions.type())),
+                    axes),
+                Collections.singletonList(tf.constant("predictions must be <= 1"))));
+
+    LossTuple<T> result =
+        LossesHelper.squeezeOrExpandDimensions(tf, tLabels, tPredictions, tSampleWeight);
+    tPredictions = result.getTarget();
+    tLabels = result.getLabels();
+    tSampleWeight = result.getSampleWeights();
+
+    if (!tPredictions.shape().isCompatibleWith(tLabels.shape()))
+      throw new IllegalArgumentException(
+          String.format(
+              "Shapes %s and %s are incompatible)",
+              tPredictions.shape().toString(), tLabels.shape().toString()));
+
+    if (topK != null) {
+      tPredictions = filterTopK(tf, tPredictions, topK);
+    }
+
+    if (classIndex != null) {
+      tLabels = tf.gather(tLabels, tf.constant(new int[] {classIndex}), tf.constant(1));
```

Comment:
I finally read the corresponding code in the Python implementation, and it has different semantics from what's here. It reads:
~~~Python
  if class_id is not None:
    y_true = y_true[..., class_id]
    y_pred = y_pred[..., class_id]
~~~
This uses `class_id` as an index into the last dimension of predictions or labels. Our Java code currently uses it as an index into dimension 1.

If you switch to the same logic as Python, I'd be a lot more comfortable than previously with the name `classId`.

Reply:
In Python, Ellipsis is used for slicing multidimensional `numpy` arrays. So in the example, ` [..., class_id],` it is pulling the `class_id` column out. `tf.gather` does the same thing. It is pulling out the `classIndex` column, and only 1 column.

Reply:
In the `!multiLabel` case, we support multiple class dimensions. `[..., class_id]` uses `class_id` as an index into the last class dimension. `tf.gather(tLabels, tf.constant(new int[] {classIndex}), tf.constant(1))` uses `classIndex` as an index into the first class dimension. (Which is what made me uncomfortable with the name `classId`.)

Reply:
You're right:
Consider this in Python:

```
y_true = np.array([[[0,1], [2,3]],[[4,5],[6,7]]])
# y_true shape = (2, 2, 2)
class_id = 1
y_true = y_true[..., class_id]
# y_true shape = (2, 2)
print(y_true)
# [[1, 3,] [5, 7]]
```

To achieve this result with `tf.gather,` I need to change the `axis` to `-1`:
```
int [][][] array = new int[][][] {{{0,1}, {2,3}},{{4,5},{6,7}}};
Operand<TInt32> iop = tf.constant(array);
// iop shape = [2, 2, 2]
int classIndex = 1;
Operand<TInt32> result =  tf.gather(iop,
              tf.constant(new int[] {classIndex}),
              tf.constant(-1));
// result shape = [2, 2, 1]

// [0, 0, 0]) 1
// [0, 1, 0]) 3
// [1, 0, 0]) 5
// [1, 1, 0]) 7
```
However, the rank of the result is too high.
If I add a `tf.squeeze` to the result, then I get:
```
result = tf.squeeze(result);
// shape = [2, 2]
// [0, 0]) 1
// [0, 1]) 3
// [1, 0]) 5
// [1, 1]) 7
```

Let me know if this looks ok.

Reply:
I think the call to `tf.squeeze` has to specify dimension `-1`:
~~~Java
result = tf.squeeze(result, Squeeze.axis(Collections.singletonList(-1L)))
~~~
Then this looks right to me. And now I'd feel fine going back to the name `classId`.